vllm/v1/stats/common.py

```diff
@@ -0,0 +1,393 @@
+      this reference (examples include per decoded token RequestOutput,
+      EngineCoreOutput, etc.).
+    """
+
+    class Type(IntEnum):
+        """See `RequestStats` for the lifecycle of a request."""
+        # Request arrived at the engine frontend.
+        ARRIVED = 0
+        # Input processed by the input processor.
+        INPUT_PROCESSED = 1
+        # Queued on the engine core.
+        QUEUED = 2
+        # Scheduled running by the scheduler.
+        RUNNING = 3
+        # Preempted by the scheduler.
+        PREEMPTED = 4
+        # Token decoded by the engine.
+        DECODED = 5
+        # Token detokenized by the detokenizer.
+        DETOKENIZED = 6
+
+    request_id: str
+
+    type: Type
+
+    # Timestamp when the update is recorded. This is used to record time
+    # intervals between events rather than wall clock time.
+    monotonic_ts_s: float = msgspec_field(
+        default_factory=lambda: time.monotonic())
+
+    ############################################################
+    # Metadata associated with the update.
+    ############################################################
+    # For input_processed.
+    # NOTE: it's fine to keep a reference to the engine request here
+    # because there will only be 1 event with this reference for a
+    # request. We need metadata (e.g. prompt tokens, sampling_param)
+    # from the request.
+    engine_request: Optional[EngineCoreRequest] = None
+
+    # For running.
+    # If the request was a newly scheduled request running for prefill.
+    new_prefill: Optional[bool] = None
+    # Number of tokens computed when scheduled to run.
+    num_computed_tokens: Optional[int] = None
+    # Number of cached tokens when scheduled to run.
+    num_cached_tokens: Optional[int] = None
+
+    # For decoded.
+    # The perfcounter timestamp for each output token.
+    token_perf_ts_ns: Optional[int] = None
+    # The number of new output tokens generated.
+    num_new_tokens: Optional[int] = None
+
+    # For both detokenized and decoded.
+    # Finished reason.
+    finish_reason: Optional[str] = None
+
+
+@dataclass
+class RequestStats:
+    """Stats associated with a request (`Request`)
+
+    A request would go through the following lifecycles upon arriving
+    at the llm engine:
+    - Arrival: when the request is first added to the llm engine.
+    - Inputs processed: when the input processor is completed.
+    - Queued: added to the waiting queue of the scheduler in the EngineCore.
+    - Running(prefill): when the request is scheduled by the scheduler
+        for prefill.
+    - [Preempted]: a request could be temporarily unscheduled by the scheduler
+        under contention of resources. This will go back to the waiting queue
+        of the scheduler, and the request will be scheduled again.
+    - Decoding: when a request is in decoding stage, and tokens are generated.
+    - Detokenized: when tokens are detokenized by the detokenizer.
```

Comment:
note: these stages are not mutually exclusive:
- `Detokenized` stage is part of decoding (and actually runs concurrently with the `EngineCore` in V1